# Tugas

## Tugas 1

Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data `spam.csv`
2. Fitur `CountVectorizer` dengan mengaktifkan **stop_words**
3. Evaluasi hasilnya

## Tugas 2

Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data `spam.csv`
2. Fitur `TF-IDF` dengan mengaktifkan **stop_words**
3. Evaluasi hasilnya dan bandingkan dengan hasil tugas 1.
4. Berikan kesimpulan fitur mana yang terbaik pada kasus data `spam.csv`

# Tugas 1

In [1]:
import pandas as pd
import numpy as np

# Read the file spam.csv into a DataFrame df using the read_csv() function.

df = pd.read_csv('spam.csv', encoding='latin-1')

# drop the unnamed columns
df = df.drop(df.iloc[:,2:], axis=1)

df.columns = ['label', 'text']

df.head()

# Encode label to a numerical variable

df['label'] = df['label'].map({'ham':0, 'spam':1})


from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer


# Split label and data.
X = df['text']
y = df['label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)

bow = CountVectorizer(stop_words='english')

X_train = bow.fit_transform(X_train)

X_test = bow.transform(X_test)

# train model 

from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score

# Inisiasi MultinomialNB
mnb = MultinomialNB()

# Fit model
mnb.fit(X_train, y_train)

# Prediksi dengan data training
y_pred_train = mnb.predict(X_train)

# Evaluasi akurasi data training
acc_train = accuracy_score(y_train, y_pred_train)

# Prediksi dengan data training
y_pred_test = mnb.predict(X_test)

# Evaluasi akurasi data training
acc_test = accuracy_score(y_test, y_pred_test)

# Print hasil evaluasi
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')


Hasil akurasi data train: 0.9946152120260264
Hasil akurasi data test: 0.9829596412556054


# Tugas 2

In [2]:
import pandas as pd
import numpy as np

# Read the file spam.csv into a DataFrame df using the read_csv() function.

df_2 = pd.read_csv('spam.csv', encoding='latin-1')

# drop the unnamed columns
df_2 = df_2.drop(df_2.iloc[:,2:], axis=1)

df_2.columns = ['label', 'text']

df_2.head()

# Encode label to a numerical variable

df_2['label'] = df_2['label'].map({'ham':0, 'spam':1})


from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer


# Split label and data.
X_2 = df_2['text']
y_2 = df_2['label']

X_2_train, X_2_test, y_2_train, y_2_test = train_test_split(X_2, y_2, test_size=0.2, random_state=50)

bow = TfidfVectorizer(stop_words='english')

X_2_train = bow.fit_transform(X_2_train)

X_2_test = bow.transform(X_2_test)

# train model 

from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score

# Inisiasi MultinomialNB
mnb_2 = MultinomialNB()

# Fit model
mnb_2.fit(X_2_train, y_2_train)

# Prediksi dengan data training
y_2_pred_train = mnb_2.predict(X_2_train)

# Evaluasi akurasi data training
acc_2_train = accuracy_score(y_2_train, y_2_pred_train)

# Prediksi dengan data training
y_2_pred_test = mnb_2.predict(X_2_test)

# Evaluasi akurasi data training
acc_2_test = accuracy_score(y_2_test, y_2_pred_test)

# Print hasil evaluasi
print(f'Hasil akurasi data train: {acc_2_train}')
print(f'Hasil akurasi data test: {acc_2_test}')


Hasil akurasi data train: 0.9842943684092439
Hasil akurasi data test: 0.9605381165919282


# Kesimpulan
Pada saat menggunakan **Count Vectorizer** dengan stop_words adalah english, maka hasil akurasi data yang dihasilkan adalah sebagai berikut:
    

> Hasil akurasi data train: 0.9946152120260264   
> Hasil akurasi data test: 0.9829596412556054

Sedangkan, ketika menggunakan **TfIdfVectorizer** dengan stop_words english, maka hasil akurasi data yang dihasilkan adalah sebagai berikut:

> Hasil akurasi data train: 0.9842943684092439   
> Hasil akurasi data test: 0.9605381165919282

Sehingga, dapat **disimpulkan** bahwasannya, hasil keakurasian datanya lebih baik ketika menggunakan **Count Vectorizer**
